### PR TITLE
Expose `database#compactRange(start, end, callback)`

### PR DESCRIPTION
**Pull request by bookchin**
This PR adds the ability to manually trigger a database compaction within a given key range for the purpose of providing better control over how and when compaction occurs.

### CODE REVIEW

**Comment by bookchin** [comment]
Travis build appears as "cancelled" - it may need a restart.

**Comment by bookchin** [comment]
After looking at Travis, it seems that all the OSX builds were cancelled (all GNU+Linux passed). I'm not familiar with how Travis does multi-job build like that or how to trigger a restart.

@ralphtheninja or halp pl0x? <3

**Comment by ralphtheninja** [comment]
I have no idea what's causing this. Lets wait a bit and restart the build. Travis fights back sometimes.

@juliangruber Any ideas on the osx builds?

**Comment by bookchin** [comment]
Woo! 🎉  Thanks for kicking it!

**Comment by vweevers** [comment]
Travis experienced major outage this week and had to cancel mac builds ([incident report](https://www.traviscistatus.com/incidents/k79mjcv403c4)).

**Comment by bookchin** [comment]
@juliangruber thanks!

In case you are interested, we needed the ability to have more control over compaction to free physical disk space at specific times instead of relying solely on leveldb's default strategy. So now that we can periodically trigger compaction according to our application specific needs, we saw moderate improvement in read performance and significantly less variance in read performance as well.

To compare, the left hand side of this test run was before the manual "flushing" - which already gave significant improvements over using "vanilla" leveldb:



Compare those results to these after performing compaction manually on the right hand side of these results:

File: test/compact-range-test.js [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  var val2 = Buffer(64).fill(1);
+  db.put(key1, val1, function() {
+    db.put(key2, val2, function() {
+      db.compactRange(key1, key2, function() {
+        db.approximateSize('0', 'z', function(err, sizeAfterPuts) {
+          db.del(key1, function() {
+            db.del(key2, function() {
+              db.compactRange(key1, key2, function() {
+                db.approximateSize('0', 'z', function(err, sizeAfterCompact) {
+                  t.notEqual(sizeAfterCompact, sizeAfterPuts);
```
Comment (ralphtheninja) [inline test/compact-range-test.js]: `sizeAfterCompact` should technically be less than `sizeAfterPuts` right?

File: package.json [hunk 2]
```diff
@@ -15,7 +15,8 @@
     "Anton Whalley <anton.whalley@nearform.com> (https://github.com/No9)",
     "Matteo Collina <matteo.collina@gmail.com> (https://github.com/mcollina)",
     "Pedro Teixeira <pedro.teixeira@gmail.com> (https://github.com/pgte)",
-    "James Halliday <mail@substack.net> (https://github.com/substack)"
+    "James Halliday <mail@substack.net> (https://github.com/substack)",
+    "Gordon Hall <gordonh@member.fsf.org> (https://github.com/bookchin)"
```
Comment (ralphtheninja) [inline package.json]: :+1: @bookchin Want to be part of the level organization as well?

File: README.md [hunk 3]
```diff
@@ -172,6 +172,16 @@ The `callback` function will be called with no arguments if the operation is suc
 
 
 --------------------------------------------------------
+<a name="leveldown_compactRange"></a>
+### leveldown#compactRange(start, end, callback)
```
Comment (ralphtheninja) [inline README.md]: Please add a link at the top of the readme pointing to this section.

**Comment by ralphtheninja** [comment]
@bookchin Very nice work :) Out of curiosity, what are you using to visualize those statistics?

File: test/compact-range-test.js (hunk 1 above)
Comment (bookchin) [inline test/compact-range-test.js]: Ah, yes, I will update that assertion

File: package.json (hunk 2 above)
Comment (bookchin) [inline package.json]: Oh! I didn't mean to imply that by that addition - force of habit (I generally encourage any contributor to "sign my yearbook"), but given that our work on @Storj makes extensive use of this project, I'd be thrilled to continue to contribute!

File: README.md (hunk 3 above)
Comment (bookchin) [inline README.md]: Will do

**Comment by bookchin** [comment]
@ralphtheninja these graphs were generated by our data scientist @pgerbes1 who would be happy to share the R scripts used to produce these. 

In the meantime, there is more info on this at https://storj.github.io/kfs/tutorial-performance-testing.html

File: package.json (hunk 2 above)
Comment (bookchin) [inline package.json]: (yes) 😊

File: README.md (hunk 3 above)
Comment (bookchin) [inline README.md]: Done!

File: package.json (hunk 2 above)
Comment (ralphtheninja) [inline package.json]: Invited. You have full access like everyone else.

**Comment by bookchin** [comment]
Thanks!

**Pull request merged**